test/integration/master/secrets_enveloping_test.go

```diff
@@ -0,0 +1,318 @@
+	// Envelope header precedes the payload
+	sealedData := response.Kvs[0].Value[len(expectedEnvelopePrefix):]
+	v, err := unSealSecretFunc(sealedData, ctx, transformerConfig)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	if !strings.Contains(string(v), secretVal) {
+		t.Errorf("expected %q after decryption, but got %q", secretVal, string(v))
+	}
+
+	// Secrets should be un-enveloped on direct reads from Kube API Server.
+	s, err := client.CoreV1().Secrets(testNamespace).Get(testSecret, metav1.GetOptions{})
+	if secretVal != string(s.Data[secretKey]) {
+		t.Errorf("expected %s from KubeAPI, but got %s", secretVal, string(s.Data[secretKey]))
+	}
+}
+
+func startKubeApiWithEncryptionOrDie(t *testing.T, storageConfig *storagebackend.Config,
+	transformerConfig string) *kubeapiservertesting.TestServer {
+	tempDir, err := ioutil.TempDir("", "secrets-encryption-test")
+	if err != nil {
+		t.Errorf("failed to create temp directory: %v", err)
+	}
+	defer os.RemoveAll(tempDir)
+
+	encryptionConfig, err := ioutil.TempFile(tempDir, "encryption-config")
+	if err != nil {
+		t.Errorf("error while creating temp file for encryption config %v", err)
+	}
+
+	if _, err := encryptionConfig.Write([]byte(transformerConfig)); err != nil {
+		t.Errorf("error while writing encryption config: %v", err)
+	}
+
+	kubeAPIOptions := []string{"--experimental-encryption-provider-config", encryptionConfig.Name()}
+	return kubeapiservertesting.StartTestServerOrDie(t, kubeAPIOptions, storageConfig)
+}
+
+func createTestNamespace(client *kubernetes.Clientset, name string) (corev1.Namespace, error) {
+	ns := corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+	}
+
+	if _, err := client.CoreV1().Namespaces().Create(&ns); err != nil {
+		return corev1.Namespace{}, fmt.Errorf("unable to create testing namespace %v", err)
+	}
+
+	return ns, nil
+}
+
+func createTestSecret(client *kubernetes.Clientset, name, namespace string) (corev1.Secret, error) {
+	secret := corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Data: map[string][]byte{
+			secretKey: []byte(secretVal),
+		},
+	}
+	if _, err := client.CoreV1().Secrets(secret.Namespace).Create(&secret); err != nil {
+		return corev1.Secret{}, fmt.Errorf("error while writing secret: %v", err)
+	}
+
+	return secret, nil
+}
+
+func readRawRecordFromETCD(kubeAPIServer *kubeapiservertesting.TestServer, path string) (*clientv3.GetResponse, error) {
+	// Reading secret directly from etcd - expect data to be enveloped and the payload encrypted.
+	etcdClient, err := createETCDClient(kubeAPIServer.ServerOpts.Etcd.StorageConfig)
+	if err != nil {
+		return nil, err
+	}
+	response, err := etcdClient.KV.Get(context.Background(), path, clientv3.WithPrefix())
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve secret from etcd %v", err)
+	}
+
+	return response, nil
+}
+
+func createETCDClient(storageConfig storagebackend.Config) (*clientv3.Client, error) {
```

Comment:
Could you search for the similar method? It looks like a helper and, perhaps we already have it, or we can extract it to utils and share with other tests.

Reply:
Extracted GetEtcdKVClient into test/integration/utils.go

There is a possibility to generalize etcd related code in test/integratin/etcd into a higher level etcd client - another PR?

Reply:
> There is a possibility to generalize etcd related code in test/integratin/etcd into a higher level etcd client - another PR?

I'd rather just move the `createETCDClient()` function there. But it's up to you.